### ISSUE TITLE
Multiple entries not outputting multiple entry points

### ISSUE THREAD

**Issue by fnky** [palmerhq/tsdx#175]
### Current Behavior

When building source with multiple entries, using `--entry src/**.ts` the build only uses the source of the first file it finds as the main entry point for ESM and CJS output.

### Desired Behavior

It would be nice to be able to specify multiple destinations for different source files from a single codebase.

### Suggested Solution

A solution could be to change the behavior of `--entry` flag, so that instead of outputting only definition files from multiple sources, it should output entry points for each specified entry file.

For example, with a folder structure like this:

```
src/
  code.ts
  ui.ts
```

Running `tsdx build --entry src/**.ts` or `--entry src/code.ts src/ui.ts` will yield the following output files:

```
dist/
  code.js
  code.cjs.production.js
  code.cjs.development.js
  code.esm.js
  ui.js
  ui.cjs.production.js
  ui.cjs.development.js
  ui.esm.js
```

Where `code.js` and `ui.js` are similar to `index.js` for single entry projects.

### Who does this impact? Who is this for?

This is useful for building applications, plugins or other types of projects, where multiple entries are required, without having to create a monolithic repo architecture with a custom build process to combine build files.

An example is the recently released Figma Plugin API, which specifies a main file for the underlying plugin code, and a ui file for browser UI part.

See an [example project using React and Webpack to create a Figma UI plugin](https://github.com/figma/plugin-samples/tree/master/react)

### Describe alternatives you've considered

- Creating a monorepo and using a custom script to merge the builds into a single folder.
- Move to a custom setup with webpack that allows for this type of configuration.

### Additional context

I was looking for ways to specify multiple entries with `--entry` as outlined in https://github.com/palmerhq/tsdx/pull/28 but it seems to only output source files for the first file it finds and the rest just outputs as definition files, not actual source files. This is a bit confusing and may be a bug as it shares the same property name as webpack (`entry`), but works differently.

**Comment by elado** [comment]
another use case is adding command line tool (`bin`) to a package in a separate `cli.ts` file.
but here, there's no need to use esm/umd/cjs builds.

`tsdx.config.js` `rollup` method returns a single entry per build format (esm/umd/cjs) but would be great to be able to also hook into the final config, so more entries can be injected.

**Comment by jaredpalmer** [comment]
Yep. This should be easy to implement just have to fool around with the cjs entry generation so it does it multiple times

**Comment by darthtrevino** [comment]
To follow up with @elado 's comment, if your code uses web-workers at all it would be nice to be able to use multiple entry points split across output files so that worker scripts are separate from everything else.

**Comment by yordis** [comment]
Related to this I guess,

How could I avoid having one single outputted file for ESM build?

```
src/
  folder/
    file.ts
  code.ts
  ui.ts
```

```
dist/
  folder/
    file.js
  code.js
  ui.js
```

I am looking for having something like Material UI folks put together.

**Comment by agilgur5** [comment]
So this seems to be a bug caused by the `rollup` config's `output.file` being set to the same thing for every entry file:
https://github.com/jaredpalmer/tsdx/blob/d621994f3eb21b6f686cccf7170ba541e72886c3/src/createRollupConfig.ts#L32-L33

I created a workaround-ish with `tsdx.config.js` to support multiple entries:
```javascript
module.exports = {
  rollup(config, options) {
    const outputDir = process.cwd() + '/dist/'
    const extension = '.' + config.output.file.split('.').slice(1).join('.')
    let filename = config.input.split('src/')[1] // remove src/
    filename = filename.split('.')[0] // remove extension, if any

    config.output.file = outputDir + filename + extension
    console.log(config.output.file)
    return config
  }
}
```
and this will work with `tsdx build --entry src/index.ts --entry src/cli.ts`, outputting `dist/index.js` and `dist/cli.js` and their respective formats (`cjs`, `esm`, etc) and type declarations.

<br>

_Buuut_ besides being very hacky, the output has some issues (possibly minor, pending your use case):
- The `index.ts` file is output as `index.js` instead of `${packageName || nameArg}.js` like it normally is (though it probably makes more sense for the case of multiple entry files to not have a default name)
- Entries in a _subdirectory_ will output what seem to be extraneous type declarations, e.g. if you had `src/cli/index.ts` (instead of `src/cli.ts`), you will get `dist/cli/index.js` output correctly, and its types _seem_ to output correctly, but then you'll also get an extraneous output of `dist/cli/src/**/*.d.ts` that just copies all the type declarations again.
Now this might be because I have some modules shared between my two entries, which causes some bugs, maybe it won't if they're not sharing modules. In any case, if all of your entries are in the same directory (e.g. `src/`) this doesn't seem to pop up

<hr>

Also #365 seems directly related to this. Also related to this issue, @yordis created a separate issue #321 for that specific use case above, which I also gave a partial workaround for in https://github.com/jaredpalmer/tsdx/issues/321#issuecomment-564784332

**Comment by agilgur5** [comment]
So I realized there's one other fairly _big_ issue with my workaround:
`tsdx` only ever outputs one CJS "entry file" (the one that splits between dev & prod builds), so while `dist/index.js` will be created, `dist/cli.js` _will not_ be created, only `dist/cli.esm.js`, `dist/cli.cjs.development.js`, `dist/cli.cjs.production.min.js` and `.d.ts` extensions will be (and any other formats you might specify).
There's no way to workaround this in `tsdx.config.js`, it has to be fixed internally. I submitted a PR (#367) to resolve these issues and make `--entry src/index.ts --entry src/cli.ts` work correctly without hacky workarounds. 

If you're not using CJS though, you don't have a need for CJS entry files, so the hacky workaround could potentially work for you.

Another issue is that `package.json` fields like `module` won't apply for more than one entry, since `module` only accepts a single file. This could be resolved by creating more `package.json`s though... but that's fairly complex of a solution. Directing your users to directly import a `.esm.js` file or just use the CJS build is the alternative there 😕

**Comment by brainkim** [comment]
Does anyone know how to do this just with raw rollup? I’m having trouble deduplicating my own files/making d.ts files show up in the right places.

**Comment by ambroseus** [comment]
@brainkim guess this would be helpful
 https://levelup.gitconnected.com/code-splitting-for-libraries-bundling-for-npm-with-rollup-1-0-2522c7437697

**Comment by kevboh** [comment]
Hello! I just ran into this as well. Would love some way to export a library that also bundles a CLI, and I'm happy to pitch in to make that happen.

**Comment by felixmosh** [comment]
Is it possible to use [`rollup-plugin-multi-input`](https://www.npmjs.com/package/rollup-plugin-multi-input) for this?

**Comment by yaquawa** [comment]
Hey guys, I'm happy with using [tsup](https://tsup.egoist.sh/#bundle-files) now 👍🏼

**Comment by mikestopcontinues** [comment]
Hey, what's the status of this? It looks like it's holding up a lot of projects, judging by the refs.

**Comment by fnky** [comment]
@mikestopcontinues It's currently part of the [v0.15.0 milestone](https://github.com/formium/tsdx/milestone/3). Not sure if it's actively worked on at the moment. I have personally moved away from TSDX for now.

**Comment by mikestopcontinues** [comment]
Thanks @fnky . I guess I'll stick with microbundle for now, but I'll keep an eye out here. :)

**Comment by flybayer** [comment]
I've switched to https://github.com/preconstruct/preconstruct

**Comment by y-nk** [comment]
can't believe it's a year without fixing. @jaredpalmer do you need some help with that?

**Comment by ImanMh** [comment]
Can somebody from TS team mentions if this is totally a bad idea or it's on the roadmap so I can make the best decision about my problem?

**Comment by mikestopcontinues** [comment]
@ImanMh At this point, I think you need to look elsewhere. Look into esbuild. It handles typescript, .d.ts files, and multiple entries.
